node/src/components/block_accumulator.rs

```diff
@@ -227,37 +243,63 @@ impl BlockAccumulator {
         &mut self,
         block_hash: BlockHash,
         maybe_era_id: Option<EraId>,
-        sender: NodeId,
+        maybe_sender: Option<NodeId>,
     ) {
-        let maybe_local_tip_era_id = self.local_tip.map(|local_tip| local_tip.era_id);
-        match (maybe_era_id, maybe_local_tip_era_id) {
-            // When the era of the item sent by this peer is known, we check it
-            // against our local tip era. If it is recent (within a number of
-            // eras of our local tip), we create an acceptor for it along with
-            // registering the peer.
-            (Some(era_id), Some(local_tip_era_id))
-                if era_id >= local_tip_era_id.saturating_sub(self.recent_era_interval) =>
-            {
-                let acceptor = self
-                    .block_acceptors
-                    .entry(block_hash)
-                    .or_insert_with(|| BlockAcceptor::new(block_hash, None));
-                acceptor.register_peer(sender);
-            }
-            // In all other cases (i.e. the item's era is not provided, the
-            // local tip doesn't have an era or the item's era is older than
-            // the local tip era by more than `recent_era_interval`), we only
-            // register the peer if there is an acceptor for the item already.
-            _ => {
-                if let Some(acceptor) = self.block_acceptors.get_mut(&block_hash) {
-                    acceptor.register_peer(sender)
+        // If the acceptor already exists, just register the peer, if applicable.
+        let entry = match self.block_acceptors.entry(block_hash) {
+            btree_map::Entry::Occupied(entry) => {
+                if let Some(sender) = maybe_sender {
+                    entry.into_mut().register_peer(sender);
                 }
+                return;
+            }
+            btree_map::Entry::Vacant(entry) => entry,
+        };
+
+        // The acceptor doesn't exist. Don't create it if the item's era is not provided, the local
+        // tip doesn't have an era or the item's era is older than the local tip era by more than
+        // `recent_era_interval`.
+        match (maybe_era_id, self.local_tip) {
+            (Some(era_id), Some(local_tip))
+                if era_id >= local_tip.era_id.saturating_sub(self.recent_era_interval) => {}
+            _ => return,
+        }
+
+        // Check that the sender isn't telling us about more blocks than expected.
+        if let Some(sender) = maybe_sender {
+            let block_timestamps = self.peer_block_timestamps.entry(sender).or_default();
+
+            // Prune the timestamps, so the count reflects only the most recently added acceptors.
+            let purge_interval = self.purge_interval;
+            let is_expired = |(_, timestamp): &(_, Timestamp)| timestamp.elapsed() > purge_interval;
+            while block_timestamps.front().map_or(false, is_expired) {
+                block_timestamps.pop_front();
+            }
```

Comment:
Can we inline the `is_expired`, making the type annotations redundant?

```suggestion
            while block_timestamps
                .front()
                .map_or(false, |(_, timestamp)| timestamp.elapsed() > purge_interval)
            {
                block_timestamps.pop_front();
            }
```